src/outputs/rcp.c

```diff
@@ -0,0 +1,1452 @@
+   * remote side
+   */
+  device->prevent_playback = 0;
+
+  s->callback_id = callback_id;
+  // tear this session down, incl free'ing it
+  rcp_session_shutdown_init(s);
+
+  return 1;
+}
+
+static int
+rcp_device_flush(struct output_device *device, int callback_id)
+{
+  DPRINTF(E_DBG, L_RCP, "device flush %ld '%s'\n", device->id, device->name);
+
+  // needed?
+  struct rcp_session *s = device->session;
+
+  s->callback_id = callback_id;
+  s->state = OUTPUT_STATE_STOPPED;
+
+  rcp_status(s);
+
+  return 1;
+}
+
+static int
+rcp_device_probe(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device probe %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+
+  return 1;
+}
+
+
+static int
+rcp_device_volume_set(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+  char cmd[4];
+  int ret;
+
+  DPRINTF(E_DBG, L_RCP, "device volume set %ld '%s' state=%d vol=%d\n", device->id, device->name, s->state, device->volume);
+
+  if (s->state != RCP_STATE_STREAMING)
+    return 0;
+
+  s->callback_id = callback_id;
+
+
+  ret = snprintf(cmd, sizeof(cmd), "%d", device->volume);
+  if (ret < 0) {
+    return 0;
+  }
+
+  rcp_send(s, RCP_STATE_VOL_SET, cmd);
+
+  return 1;
+}
+
+static void
+rcp_device_cb_set(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+
+  DPRINTF(E_DBG, L_RCP, "device cb %ld '%s'  callback %d\n", device->id, device->name, callback_id);
+
+  s->callback_id = callback_id;
+}
+
+static void
+rcp_device_add(const char *device_name, const char *address, int port)
+{
+  struct output_device *device;
+  uint64_t id;
+  int ret;
+
+  /* possible if this was mdns/avahi that there is a timeout/cached entry is no
+   * longer resolvable;  ie avahi caches a roku name, the roku is pwoered down,
+   * cache entry still exists but now no address
+   */ 
+  if (!address)
+    {
+      DPRINTF(E_DBG, L_RCP, "Attempt to add '%s' without address (possible mdns timeout)\n", device_name ? device_name : "<no name>");
+      return;
+    }
+
+  id = djb_hash(device_name, strlen(device_name));
+  if (!id)
+    {
+      DPRINTF(E_LOG, L_RCP, "Could not hash RCP name for id (%s)\n", device_name);
+      return;
+    }
+
+  CHECK_NULL(L_RCP, device = calloc(1, sizeof(struct output_device)));
+
+  device->id = id;
+  device->name = strdup(device_name);
+  device->type = OUTPUT_TYPE_RCP;
+  device->type_name = outputs_name(device->type);
+  // RCP/Roku Soundbridges only support ipv4
+  device->v4_address = strdup(address);
+  device->v4_port = port;
+  DPRINTF(E_INFO, L_RCP, "Adding RCP output device '%s' at '%s'\n", device_name, address);
+
+  ret = player_device_add(device);
+  if (ret < 0)
+    outputs_device_free(device);
+}
+
+static void
+rcp_device_remove(const char *device_name, const char *address, int port)
```

Comment:
Need to add that this function is peculiar. The address and port check don't seem to serve a purpose, the djb_hash error check shouldn't be there since djb_hash cannot fail (same in rcp_device_add), and the logging of "Removing..." is already covered by outputs_device_remove().

Duplicate code could also be avoided by instead doing like in e.g. airplay.c.

Reply:
will clean this up.  FWIW, the `djb_hash()` check is cloned from `cast.c`